File: MD-GIS/MD-GIS/App_Code/Ship.vb
Public Class Ship

    Public Property DimensionToBow As Integer
    Public Property DimensionToStern As Integer
    Public Property DimensionToPort As Integer
    Public Property DimensionToStarboard As Integer
    Public Property ETA As String
    Public Property RepeatIndicator As Integer
    Public Property RateOfTurn As Double
    Public Property Draught As Double
    '------------ Old Data ---------------
    Public Property Id As String
    Public Property MMSI As String
    Public Property IMONumber As String
    Public Property VesselName As String
    Public Property AISversion As Integer
    Public Property CallSign As String
    Public Property DataDate As String
    Public Property DataTime As String

    Public Property Latitude As Double
    Public Property Longitude As Double
    Public Property LastLatitude As Double
    Public Property LastLongitude As Double

    Public Property SpeedOverGround As Double
    Public Property CourseOverGround As Double
    Public Property ShipTypeId As String
    Public Property ShipTypeName As String
    Public Property ShipTypeColor As String
    Public Property Flag As String
    Public Property Origin As String
    Public Property Destination As String
    Public Property NoOfSattelites As Integer
    Public Property Altitude As Double
    Public Property MagneticVariation As Double
    Public Property GPSIndicator
    Public Property GoeidalSeparation
    '------------ Old Data ---------------
    Public Property NavigationStatusId As String = ""
    Public Property NavigationStatusName As String = ""
    Public Property TrueHeading As Integer
    Public Property AISchannel As String

    Public Property IsBaseStation As Boolean = False
    Public Property baseStationMMSI As String = ""

    Public Property OptimizedUI As UI

    Public Class UI

        Public Property DimensionToBow As String
        Public Property DimensionToStern As String
        Public Property DimensionToPort As String
        Public Property DimensionToStarboard As String
        Public Property ETA As String
        Public Property RepeatIndicator As Integer
        Public Property RateOfTurn As String
        Public Property Draught As String
        Public Property Id As String
        Public Property MMSI As String
        Public Property IMONumber As String
        Public Property VesselName As String
        Public Property AISversion As String
        Public Property CallSign As String
        Public Property DataDate As String
        Public Property DataTime As String

        'Public Property Updated As String = ""
        Public Property Position As String

        Public Property Latitude As Double
        Public Property Longitude As Double
        Public Property LastLatitude As Double
        Public Property LastLongitude As Double

        Public Property SpeedOverGround As String = ""
        Public Property CourseOverGround As String = ""

        Public Property ShipTypeId As String = ""
        Public Property ShipTypeName As String = ""
        Public Property ShipTypeColor As String = ""

        Public Property Flag As String = ""
        Public Property FlagCode As String = "" '-------------- เพิ่มเข้ามาแก้ ----------------

        Public Property Origin As String = ""
        Public Property Destination As String = ""
        Public Property NoOfSattelites As String = ""
        Public Property Altitude As String = ""
        Public Property MagneticVariation As String = ""
        Public Property GPSIndicator As String = ""
        Public Property GoeidalSeparation As String = ""

        Public Property NavigationStatusId As String = ""
        Public Property NavigationStatusName As String = ""
        Public Property TrueHeading As String = ""
        Public Property AISchannel As String = ""

        Public Property IsBaseStation As Boolean = False
        Public Property baseStationMMSI As String = ""

        Public Property HasPicture As Boolean = False

    End Class

    Public Sub OptimizeUI()
        Dim UI As New UI
        Dim BL As New MDBL

        On Error Resume Next
        UI.MMSI = MMSI.ToString
        UI.VesselName = VesselName.ToString
        UI.DataDate = DataDate.ToString
        UI.DataTime = DataTime.ToString

        '------------------------------Updated----------------------------
        'UI.Updated = ""

        'If Not IsNothing(DataDate) AndAlso DataDate.Length = 6 Then
        '    Dim _date As DateTime = DateTime.ParseExact(DataDate, "ddMMyy", Globalization.CultureInfo.CurrentCulture)
        '    UI.Updated &= _date.ToString("dd MMM yyyy") & " "
        'End If
        'If Not IsNothing(DataTime) AndAlso DataTime.ToString.Split(".").Length > 0 Then
        '    Dim _time As String = DataTime.ToString.Split(".")(0)
        '    If _time.Length >= 6 Then
        '        UI.Updated &= _time.Substring(0, 2) & ":" & _time.Substring(2, 2)
        '    End If
        'End If
        'UI.Updated = Trim(UI.Updated)
        'If UI.Updated.Length > 6 Then
        '    Dim TMP As DateTime = DateTime.ParseExact(UI.Updated, "dd MMM yyyy hh:mm", Globalization.CultureInfo.CurrentCulture)
        '    UI.Updated &= vbLf & ReportFriendlyTime(DateDiff(DateInterval.Minute, TMP, Now))
        'End If

        UI.IMONumber = IMONumber.ToString
        UI.CallSign = CallSign.ToString

        UI.Latitude = Latitude
        UI.Longitude = Longitude
        UI.LastLatitude = LastLatitude
        UI.LastLongitude = LastLongitude

        UI.Position = Val(FormatNumber(Latitude, 5)) & vbLf & Val(FormatNumber(Longitude, 5))
        UI.ShipTypeId = ShipTypeId.ToString
        UI.ShipTypeName = ShipTypeName.ToString
        UI.ShipTypeColor = ShipTypeColor.ToString

        UI.Flag = Flag.ToString
        UI.FlagCode = BL.GetShipFlagCode(MMSI.ToString)

        UI.Origin = Origin.ToString
        UI.Destination = Destination.ToString

        If IsNumeric(SpeedOverGround) Then
            If Math.Ceiling(SpeedOverGround) > 1 Then
                UI.SpeedOverGround = SpeedOverGround & " knot(s)"
            Else
                UI.SpeedOverGround = SpeedOverGround & " knot"
            End If
        End If

        If IsNumeric(CourseOverGround) Then UI.CourseOverGround = CourseOverGround & "°"

        If IsNumeric(DimensionToBow) Then UI.DimensionToBow = DimensionToBow & " m."
        If IsNumeric(DimensionToStern) Then UI.DimensionToStern = DimensionToStern & " m."
        If IsNumeric(DimensionToPort) Then UI.DimensionToPort = DimensionToPort & " m."
        If IsNumeric(DimensionToStarboard) Then UI.DimensionToStarboard = DimensionToStarboard & " m."

        UI.ETA = ETAToString(ETA.ToString)
        UI.RepeatIndicator = RepeatIndicator.ToString
        UI.RateOfTurn = RateOfTurn.ToString
        UI.NoOfSattelites = NoOfSattelites.ToString
        UI.Altitude = Altitude.ToString
        UI.MagneticVariation = MagneticVariation.ToString
        UI.GPSIndicator = GPSIndicator.ToString
        UI.GoeidalSeparation = GoeidalSeparation.ToString
        UI.NavigationStatusId = NavigationStatusId.ToString
        UI.NavigationStatusName = NavigationStatusName.ToString
        UI.TrueHeading = TrueHeading.ToString
        UI.AISchannel = AISchannel.ToString
        If IsNumeric(Draught) Then UI.Draught = Draught & " m."

        UI.IsBaseStation = IsBaseStation
        UI.baseStationMMSI = baseStationMMSI.ToString

        '---------- Check Picture -----------
        UI.HasPicture = (New MDBL).IsShipHasPicture(UI.mmsi)

        OptimizedUI = UI
    End Sub



    Private Function ETAToString(ByVal ETA As String) As String
        'MMddHHmm
        Try
            Dim C As New Converter
            Dim D As DateTime = C.StringToDate(ETA, "MMddHHmm")
            Return D.ToString("dd-MMM HH:mm")
        Catch ex As Exception
            Return ETA
        End Try
    End Function

End Class

File: MD-GIS/MD-GIS/js/ship.js
class Ship {
    constructor(Id, MMSI, IMONumber, ShipTypeName, VesselName, marker) {

        // Default Required Field
        this.Id = Id;
        this.MMSI = MMSI;
        this.IMONumber = IMONumber.trim();
        this.ShipTypeName = ShipTypeName.trim();
        this.VesselName = VesselName.trim();
        this.marker = marker;

        // Demo
        this.accessRestrictZone = false;
        this.accessZoneFirstTime = false;
        this.accessZoneName = '';
        
        // Extent Field
        this.AISversion = "";
        this.CallSign = "";
        this.DataDate = "";
        this.DataTime = "";

        this.Latitude = "";
        this.Longitude = "";
        this.LastLatitude = "";
        this.LastLongitude = "";

        this.SpeedOverGround = "";
        this.CourseOverGround;
        
        this.ShipTypeId = "";
        this.ShipTypeColor = "";
        this.Flag = "";
        this.Origin = "";
        this.Destination = "";
        this.NoOfSattelites = "";
        this.Altitude = "";
        this.MagneticVariation = "";
        this.GPSIndicator = "";
        this.GoeidalSeparation = "";
        this.NavigationStatusId = "";
        this.NavigationStatusName = "";
        this.TrueHeading = "";
        this.AISchannel = "";

        this.DimensionToBow = "";
        this.DimensionToStern = "";
        this.DimensionToPort = "";
        this.DimensionToStarboard = "";
        this.ETA = "";
        this.RepeatIndicator = "";
        this.RateOfTurn = "";
        this.Draught = "";

        this.IsBaseStation = false;
        this.baseStationMMSI = "";
    }

    updateExtra(data) {


        this.IMONumber = data.IMONumber.trim();
        this.ShipTypeName = data.ShipTypeName.trim();
        this.VesselName = data.VesselName.trim();

        // Extent Field
        this.AISversion = data.AISversion;
        this.CallSign = data.CallSign;
        this.DataDate = data.DataDate;
        this.DataTime = data.DataTime;

        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.LastLatitude = data.LastLatitude;
        this.LastLongitude = data.LastLongitude;

        this.SpeedOverGround = data.SpeedOverGround;
        this.CourseOverGround = data.CourseOverGround;

        this.ShipTypeId = data.ShipTypeId;
        this.ShipTypeColor = data.ShipTypeColor;
        this.Flag = data.Flag;
        this.Origin = data.Origin;
        this.Destination = data.Destination;
        this.NoOfSattelites = data.NoOfSattelites;
        this.Altitude = data.Altitude;
        this.MagneticVariation = data.MagneticVariation;
        this.GPSIndicator = data.GPSIndicator;
        this.GoeidalSeparation = data.GoeidalSeparation;
        this.NavigationStatusId = data.NavigationStatusId;
        this.NavigationStatusName = data.NavigationStatusName;
        this.TrueHeading = data.TrueHeading;
        this.AISchannel = data.AISchannel;

        this.DimensionToBow = data.DimensionToBow;
        this.DimensionToStern = data.DimensionToStern;
        this.DimensionToPort = data.DimensionToPort;
        this.DimensionToStarboard = data.DimensionToStarboard;
        this.ETA = data.ETA;
        this.RepeatIndicator = data.RepeatIndicator;
        this.RateOfTurn = data.RateOfTurn;
        this.Draught = data.Draught;

        this.IsBaseStation = data.IsBaseStation;
        this.baseStationMMSI = data.baseStationMMSI;
    }

    getLastPosition() {

    }

    setVisible(visible, map) {
        var target = map;
        if (!visible) target = null;
        this.marker.setMap(target);
    }

    getVisible() {
        return !(this.marker.getMap() == null)
    }

    isMoving() {
        try {

            if (!this.SpeedOverGround) return false;
            if (this.NavigationStatusName) {
                if (this.NavigationStatusName.toString().toLowerCase().indexOf('moore') > -1) return false;
                if (this.NavigationStatusName.toString().toLowerCase().indexOf('anchor') > -1) return false;
            }
            return true;                
        } catch (e) {
            return true;
        }       
    }

}

class Ships {
    constructor() {
        this.list = [];
    }

    add(ship) {
        var MMSI = ship.MMSI;
        if (!this.list[MMSI]) {
            this.list[MMSI] = ship;
        }
        return this;
    }

    remove(ship) {

        // Get Removed Index
        if (ship.marker) ship.marker.setMap(null);
        delete ship.marker;
        if (ship.directionLine) ship.directionLine.setMap(null);
        delete ship.directionLine;


        var target = ship.MMSI;
        var list = [];
        for (var MMSI in this.list) {
            if (MMSI != target) {
                list[MMSI] = this.list[MMSI];
            }
        }
        this.list = list;
        return this; 
    }

    indexOf(ship) {
        var i = -1;
        if (!ship) return i;
        for (var MMSI in this.list) {
            i += 1;
            if (ship.MMSI == MMSI) return i;
        }
        return -1;
    }

    removeByMMSI(MMSI) {
        var ship = this.list[MMSI];
        if (!ship) return this;
        return this.remove(ship);
    }

    byMMSI(MMSI) {
        return this.list[MMSI];
    }

    byType(ShipTypeName) {
        var result = [];
        for (var MMSI in this.list) {
            var ship = this.list[MMSI];
            if (ship.ShipTypeName == ShipTypeName)
                result.push(ship);
        }
        return result;
    }

    byVisibility(visible) {
        var result = [];
        for (var MMSI in this.list) {
            var ship = this.list[MMSI];
            if (ship.getVisible() == visible)
                result.push(ship);
        }   
        return result;        
    }
}